File: fd-proxy.js
// football-data.org 로컬 CORS 프록시
// -----------------------------------------------------------
// football-data.org는 서버용 API라 브라우저에서 직접 부르면 CORS로 막힙니다.
// 이 스크립트를 내 PC에서 띄우면 브라우저는 localhost를 부르고,
// 실제 요청은 이 프로그램이 대신 보냅니다. 토큰은 내 PC 밖으로 나가지 않습니다.
//
// 실행:  node fd-proxy.js
// 확인:  브라우저 주소창에 http://127.0.0.1:8787/__ping
// 분석기 "프록시 주소" 칸에:  http://127.0.0.1:8787
// 포트 바꾸기:  PORT=9000 node fd-proxy.js   (윈도우: set PORT=9000 && node fd-proxy.js)
// 끄기:  이 창에서 Ctrl+C
// -----------------------------------------------------------
const http = require('http');
const https = require('https');

const PORT = Number(process.env.PORT || 8787);
const UPSTREAM = 'api.football-data.org';

const CORS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'X-Auth-Token, Content-Type',
  'Access-Control-Allow-Methods': 'GET, OPTIONS',
  'Access-Control-Max-Age': '86400'
};

function handler(req, res) {
  // 브라우저는 X-Auth-Token 같은 커스텀 헤더 때문에 먼저 OPTIONS를 보냅니다.
  if (req.method === 'OPTIONS') { res.writeHead(204, CORS); return res.end(); }

  // 프록시가 살아 있는지 확인하는 용도. 토큰이 필요 없습니다.
  if (req.url === '/__ping') {
    res.writeHead(200, { ...CORS, 'Content-Type': 'application/json; charset=utf-8' });
    return res.end(JSON.stringify({ ok: true, proxy: 'fd-proxy', upstream: UPSTREAM, port: PORT }));
  }

  if (req.method !== 'GET') {
    res.writeHead(405, { ...CORS, 'Content-Type': 'application/json' });
    return res.end(JSON.stringify({ message: 'GET만 지원합니다.' }));
  }

  const token = req.headers['x-auth-token'] || '';
  const started = Date.now();
  const upstream = https.request({
    hostname: UPSTREAM, path: req.url, method: 'GET',
    headers: { 'X-Auth-Token': token, 'Accept': 'application/json' }
  }, r => {
    console.log(`  ${r.statusCode}  ${req.url}  (${Date.now() - started}ms)`);
    res.writeHead(r.statusCode || 502, { ...CORS, 'Content-Type': 'application/json; charset=utf-8' });
    r.pipe(res);
  });

  upstream.on('error', e => {
    console.log(`  실패  ${req.url}  ${e.message}`);
    res.writeHead(502, { ...CORS, 'Content-Type': 'application/json' });
    res.end(JSON.stringify({ message: 'upstream 연결 실패: ' + e.message }));
  });
  upstream.end();
}

// localhost 가 ::1(IPv6)로 먼저 해석되는 PC가 있습니다. IPv4에만 바인딩하면
// 그런 환경에서 연결이 실패하므로 두 주소 모두에서 받습니다.
let up = 0, done = 0;
const hosts = ['127.0.0.1', '::1'];
hosts.forEach(host => {
  const s = http.createServer(handler);
  s.on('error', e => {
    done++;
    if (e.code === 'EADDRINUSE') {
      console.error(`\n[오류] 포트 ${PORT}을 이미 다른 프로그램이 쓰고 있습니다.`);
      console.error(`       다른 포트로: PORT=9000 node fd-proxy.js`);
      console.error(`       (윈도우) set PORT=9000 && node fd-proxy.js`);
      process.exit(1);
    }
    // IPv6를 못 쓰는 PC는 조용히 넘어갑니다 (IPv4로 충분합니다).
    if (!['EAFNOSUPPORT', 'EADDRNOTAVAIL', 'EINVAL'].includes(e.code))
      console.error(`  ${host} 바인딩 실패: ${e.code}`);
    finish();
  });
  s.listen(PORT, host, () => { up++; done++; finish(); });
});

function finish() {
  if (done < hosts.length) return;
  if (!up) { console.error('\n[오류] 어떤 주소에도 바인딩하지 못했습니다.'); process.exit(1); }
  console.log('football-data.org 프록시 실행 중');
  console.log(`  http://127.0.0.1:${PORT}  ->  https://${UPSTREAM}`);
  console.log(`  살아있는지 확인:  http://127.0.0.1:${PORT}/__ping`);
  console.log('  분석기의 "프록시 주소" 칸에 위 주소를 넣으세요. 끄려면 Ctrl+C.\n');
}

File: fd-proxy.py
# football-data.org 로컬 CORS 프록시 (파이썬 표준 라이브러리만 사용)
# -----------------------------------------------------------
# football-data.org는 서버용 API라 브라우저에서 직접 부르면 CORS로 막힙니다.
# 이 스크립트를 내 PC에서 띄우면 브라우저는 localhost를 부르고,
# 실제 요청은 이 프로그램이 대신 보냅니다. 토큰은 내 PC 밖으로 나가지 않습니다.
#
# 실행:  python fd-proxy.py        (윈도우에서는 py fd-proxy.py)
# 확인:  브라우저 주소창에 http://127.0.0.1:8787/__ping
# 분석기 "프록시 주소" 칸에:  http://127.0.0.1:8787
# 포트 바꾸기:  PORT=9000 python fd-proxy.py
# 끄기:  이 창에서 Ctrl+C
# -----------------------------------------------------------
import http.server, urllib.request, urllib.error, json, os, socket, sys, threading, time

PORT = int(os.environ.get('PORT', 8787))
UPSTREAM = 'https://api.football-data.org'
UPSTREAM_HOST = 'api.football-data.org'

class Handler(http.server.BaseHTTPRequestHandler):
    protocol_version = 'HTTP/1.1'

    def _cors(self):
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Headers', 'X-Auth-Token, Content-Type')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Max-Age', '86400')

    def _send(self, code, body):
        self.send_response(code)
        self._cors()
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    # 브라우저는 X-Auth-Token 같은 커스텀 헤더 때문에 먼저 OPTIONS를 보냅니다.
    def do_OPTIONS(self):
        self.send_response(204); self._cors()
        self.send_header('Content-Length', '0'); self.end_headers()

    def do_GET(self):
        # 프록시가 살아 있는지 확인하는 용도. 토큰이 필요 없습니다.
        if self.path == '/__ping':
            return self._send(200, json.dumps(
                {'ok': True, 'proxy': 'fd-proxy', 'upstream': UPSTREAM_HOST, 'port': PORT}).encode())

        started = time.time()
        req = urllib.request.Request(
            UPSTREAM + self.path,
            headers={'X-Auth-Token': self.headers.get('X-Auth-Token', ''),
                     'Accept': 'application/json'})
        try:
            with urllib.request.urlopen(req, timeout=30) as r:
                body, code = r.read(), r.status
        except urllib.error.HTTPError as e:
            body, code = e.read(), e.code
        except Exception as e:
            body, code = json.dumps({'message': 'upstream 연결 실패: %s' % e}).encode(), 502
        print('  %s  %s  (%dms)' % (code, self.path, (time.time() - started) * 1000))
        self._send(code, body)

    def log_message(self, *args):
        pass

class V6Server(http.server.ThreadingHTTPServer):
    address_family = socket.AF_INET6

servers = []
try:
    servers.append(http.server.ThreadingHTTPServer(('127.0.0.1', PORT), Handler))
except OSError as e:
    print('\n[오류] 포트 %d 바인딩 실패: %s' % (PORT, e))
    print('       이미 다른 프로그램이 쓰고 있다면 다른 포트로: PORT=9000 python fd-proxy.py')
    sys.exit(1)

# localhost 가 ::1(IPv6)로 먼저 해석되는 PC가 있습니다. IPv4에만 바인딩하면
# 그런 환경에서 연결이 실패하므로 두 주소 모두에서 받습니다.
if socket.has_ipv6:
    try:
        servers.append(V6Server(('::1', PORT), Handler))
    except OSError:
        pass   # IPv6를 못 쓰는 PC는 IPv4로 충분합니다

print('football-data.org 프록시 실행 중')
print('  http://127.0.0.1:%d  ->  %s' % (PORT, UPSTREAM))
print('  살아있는지 확인:  http://127.0.0.1:%d/__ping' % PORT)
print('  분석기의 "프록시 주소" 칸에 위 주소를 넣으세요. 끄려면 Ctrl+C.\n')

for s in servers[1:]:
    threading.Thread(target=s.serve_forever, daemon=True).start()
try:
    servers[0].serve_forever()
except KeyboardInterrupt:
    print('\n프록시를 종료했습니다.')
